oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OAuth2IntrospectionAuthenticationProvider.java

```diff
@@ -0,0 +1,265 @@
+		Map<String, Object> claims = convertClaimsSet(response);
+		Instant iat = (Instant) claims.get(ISSUED_AT);
+		Instant exp = (Instant) claims.get(EXPIRES_AT);
+
+		// construct token
+		OAuth2AccessToken token  = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
+				bearer.getToken(), iat, exp);
+		Collection<GrantedAuthority> authorities = extractAuthorities(claims);
+		AbstractAuthenticationToken result =
+				new OAuth2IntrospectionAuthenticationToken(token, claims, authorities);
+		result.setDetails(bearer.getDetails());
+		return result;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private TokenIntrospectionSuccessResponse introspect(String token) {
+		HttpHeaders headers = new HttpHeaders();
+		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON_UTF8));
+		RequestEntity<String> requestEntity =
+				new RequestEntity<>("token=" + token, headers, HttpMethod.POST, this.introspectionUri);
+		TokenIntrospectionSuccessResponse response = asSuccess(parse(adapt(request(requestEntity))));
+
+		// relying solely on the authorization server to validate this token (not checking 'exp', for example)
+		if (!response.isActive()) {
+			throw new OAuth2AuthenticationException(
+					invalidToken("Provided token [" + token + "] isn't active"));
+		}
+		return response;
+	}
+
+	private ResponseEntity<String> request(RequestEntity<String> requestEntity) {
+		try {
+			return this.restOperations.exchange(requestEntity, String.class);
+		} catch (Exception ex) {
+			throw new OAuth2AuthenticationException(
+					invalidToken(ex.getMessage()), ex);
+		}
+	}
+
+	private HTTPResponse adapt(ResponseEntity<String> responseEntity) {
+		HTTPResponse response = new HTTPResponse(responseEntity.getStatusCodeValue());
+		response.setHeader(HttpHeaders.CONTENT_TYPE, responseEntity.getHeaders().getContentType().toString());
+		response.setContent(responseEntity.getBody());
+
+		if (response.getStatusCode() != HTTPResponse.SC_OK) {
+			throw new OAuth2AuthenticationException(
+					invalidToken("Introspection endpoint responded with " + response.getStatusCode()));
+		}
+		return response;
+	}
+
+	private TokenIntrospectionResponse parse(HTTPResponse response) {
+		try {
+			return TokenIntrospectionResponse.parse(response);
+		} catch (Exception ex) {
+			throw new OAuth2AuthenticationException(
+					invalidToken(ex.getMessage()), ex);
+		}
+	}
+
+	private TokenIntrospectionSuccessResponse asSuccess(TokenIntrospectionResponse introspectionResponse) {
+		if (!introspectionResponse.indicatesSuccess()) {
+			throw new OAuth2AuthenticationException(invalidToken("Token introspection failed"));
+		}
+		return (TokenIntrospectionSuccessResponse) introspectionResponse;
+	}
+
+	private Map<String, Object> convertClaimsSet(TokenIntrospectionSuccessResponse response) {
```

Comment:
They will not be `null` since this method simply updates the existing `Map`.

Nevertheless, agreed that it's better to have a test that confirms that all claims are correctly exposed as attributes, and I'll update the tests accordingly.

Reply:
> They will not be `null` since this method simply updates the existing `Map`

Yes, somehow I missed this...`Map<String, Object> claims = response.toJSONObject()`